cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,352 @@
+        )
+
+    def _value_equality_values_(self) -> Any:
+        # `isinstance` is used to ensure the a gate type and gate instance is not compared.
+        return isinstance(self.gate, raw_types.Gate), self.gate, self.name, self.description
+
+
+@value.value_equality()
+class Gateset:
+    """Gatesets represent a collection of `cirq.GateFamily` objects.
+
+    Gatesets are useful for
+        a) Describing the set of allowed gates in a human readable format
+        b) Validating a given gate / optree against the set of allowed gates
+
+    Gatesets rely on the underlying `cirq.GateFamily` for both description and
+    validation purposes.
+    """
+
+    def __init__(
+        self,
+        *gates: Union[Type[raw_types.Gate], raw_types.Gate, GateFamily],
+        name: Optional[str] = None,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> None:
+        """Init Gateset.
+
+        Accepts a list of gates, each of which should be either
+            a) `cirq.Gate` subclass
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            *gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
+                validated by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, `cirq.GlobalPhaseOperation` is accepted.
+        """
+        self._name = name
+        self._gates = frozenset(
+            g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates
+        )
+        self._unroll_circuit_op = unroll_circuit_op
+        self._accept_global_phase = accept_global_phase
+        self._instance_gate_families: Dict[raw_types.Gate, GateFamily] = {}
+        self._type_gate_families: Dict[Type[raw_types.Gate], GateFamily] = {}
+        self._custom_gate_families: List[GateFamily] = []
+        for g in self._gates:
+            if type(g) == GateFamily:
+                if isinstance(g.gate, raw_types.Gate):
+                    self._instance_gate_families[g.gate] = g
+                else:
+                    self._type_gate_families[g.gate] = g
+            else:
+                self._custom_gate_families.append(g)
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def with_params(
+        self,
+        *,
+        name: Optional[str] = None,
+        unroll_circuit_op: Optional[bool] = None,
+        accept_global_phase: Optional[bool] = None,
```

Comment:
Do we need this new `with_params` as part of the PR ? Can we not just use `**kwargs`  to cut down on redocumenting how to build the object twice ?

Reply:
Yes -- this is used in tests and we can't use **kwargs because we need to use `self.<value>` if the provided `<value>` is None.